### PR TITLE
[compliance] Improve check status report in security-agent status

### PR DESCRIPTION
**Pull request by xornivore**
### What does this PR do?

- Augments existing compliance status (based on runner-stats) with check-specific information including result and data KV map reported to the backend.
- Exports an expvar from compliance agent containing status for all running as well as failed to initialize checks.

### Motivation

Improved usability for troubleshooting compliance checks in security agent.

### Additional Notes

Example of what status changes to expect with this PR:

```
...
  cis-docker-1.2.0-3.4: Ensure that docker.socket file permissions are set to 644 or more restrictive (Scored)
  ------------------------------------------------------------------------------------------------------------
    Framework: cis-docker (1.2.0)
    Source: /etc/datadog-agent/compliance.d/cis-docker.yaml
    Configuration: [OK]

    Report:
      Result: [PASSED]
      Data:
        file.group: root
        file.path: /usr/lib/systemd/system/docker.socket
        file.permissions: 420
        file.user: root

    Total Runs: 17
    Average Execution Time : 0s
    Last Execution Date : 2020-08-12 03:10:57.000000 UTC
    Last Successful Execution Date : 2020-08-12 03:10:57.000000 UTC
...
  cis-docker-1.2.0-3.14: Ensure that the Docker server certificate key file permissions are set to 400 (Scored)
  -------------------------------------------------------------------------------------------------------------
    Framework: cis-docker (1.2.0)
    Source: /etc/datadog-agent/compliance.d/cis-docker.yaml
    Configuration: [OK]

    Report:
      Result: [ERROR]
      Data:
        error: failed to resolve path: empty path from process.flag("dockerd", "--tlskey")

    Total Runs: 1
    Average Execution Time : 3ms
    Last Execution Date : 2020-08-12 03:02:37.000000 UTC
    Last Successful Execution Date : Never
    Error: failed to resolve path: empty path from process.flag("dockerd", "--tlskey")
      No traceback

  cis-docker-1.2.0-3.15: Ensure that the Docker socket file ownership is set to root:docker (Scored)
  --------------------------------------------------------------------------------------------------
    Framework: cis-docker (1.2.0)
    Source: /etc/datadog-agent/compliance.d/cis-docker.yaml
    Configuration: [OK]
    Check has not run yet

```

### Describe your test plan

Configure various checks and validate that `security-agent status` provides correct information about check status.

### CODE REVIEW

**Review by vboulineau — APPROVED** [review]

File: pkg/compliance/checks/builder.go [hunk 1]
```diff
@@ -219,6 +220,8 @@ type builder struct {
 	dockerClient env.DockerClient
 	auditClient  env.AuditClient
 	kubeClient   env.KubeClient
+
+	status *status
```
Comment (ahmed-mez) [inline pkg/compliance/checks/builder.go:224]: could be named `checkStatus`, the same way we have `checkInterval` above

**Review by ahmed-mez — APPROVED** [review]

File: pkg/compliance/agent/agent.go [hunk 2]
```diff
@@ -92,19 +95,43 @@ func RunChecksFromFile(reporter event.Reporter, file string, options ...checks.B
...
 	return a.buildChecks(onCheck)
 }
 
-func runCheck(check check.Check) error {
-	log.Infof("%s: Running check: %s [version=%s]", check.ID(), check.String(), check.Version())
-	err := check.Run()
+func runCheck(rule *compliance.Rule, check compliance.Check, err error) bool {
+	if err != nil {
+		log.Infof("%s: Not running check: %v", rule.ID, err)
+		return true
```
Comment (remeh) [inline pkg/compliance/agent/agent.go:126]: Sanity check: we don't want to return false here?

File: pkg/compliance/agent/agent.go [hunk 3]
```diff
@@ -92,19 +95,43 @@ func RunChecksFromFile(reporter event.Reporter, file string, options ...checks.B
...
+	}
+
+	log.Infof("%s: Running check: %s [version=%s]", rule.ID, check.String(), check.Version())
+	err = check.Run()
 	if err != nil {
 		log.Errorf("%s: Check failed: %v", check.ID(), err)
```
Comment (remeh) [inline pkg/compliance/agent/agent.go:132]: Sanity check: we don't want to return false here?

**Review by remeh — APPROVED** [review]
Made a few comments but LGTM.

File: pkg/compliance/agent/agent.go (hunk 2 above)
Comment (xornivore) [reply pkg/compliance/agent/agent.go:126]: No, because that would stop enumerating other checks.

File: pkg/compliance/agent/agent.go (hunk 3 above)
Comment (xornivore) [reply pkg/compliance/agent/agent.go:132]: Same here - we want to continue enumerating other checks.

File: pkg/compliance/checks/status.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	s.ruleIDs = append(s.ruleIDs, checkStatus.RuleID)
+	s.checks[checkStatus.RuleID] = checkStatus
+}
+
+func (s *status) updateCheck(ruleID string, event *event.Event) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	stats, ok := s.checks[ruleID]
+	if !ok {
```
[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/compliance/checks/status.go:42]: Will do thank you.

File: pkg/compliance/checks/status.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/DataDog/datadog-agent/pkg/compliance"
+	"github.com/DataDog/datadog-agent/pkg/compliance/event"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// status maintains status updates for all configured checks
+type status struct {
+	ruleIDs []string
+	checks  map[string]*compliance.CheckStatus
+	mu      sync.RWMutex
```
[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/compliance/checks/status.go:20]: @remeh mind elaborating on this comment - do you mean embedding sync.RWMutex?
```
type status struct {
	sync.RWMutex
...
}
```

[the remark this answers is not in the document]
Comment (remeh) [reply pkg/compliance/checks/status.go:20]: Yes

[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/compliance/checks/status.go]: Updated thank you

File: pkg/compliance/checks/status.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (xornivore) [reply pkg/compliance/checks/status.go:42]: Added the check.

**Pull request merged**